include/zephyr/ipc/pbuf.h

```diff
@@ -0,0 +1,199 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_IPC_PBUF_H_
+#define ZEPHYR_INCLUDE_IPC_PBUF_H_
+
+#include <zephyr/cache.h>
+#include <zephyr/devicetree.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief Scure packed buffer API
+ * @ingroup kernel_apis
+ * @{
+ */
+
+/** @brief Size of packet length field. */
+#define PBUF_PACKET_LEN_SZ sizeof(uint32_t)
+
+/* Amount of data that is left unused to distinguish between empty and full. */
+#define _IDX_SIZE sizeof(uint32_t)
+
+/* Minimal length of the data field in the buffer to store the smalest packet
+ * possible.
+ * (+1) for at least one byte of data.
+ * (+_IDX_SIZE) to distinguish buffer full and buffer empty.
+ * Rounded up to keep wr/rd indexes pointing to aligned address.
+ */
+#define _MIN_DATA_LEN ROUND_UP(PBUF_PACKET_LEN_SZ + 1 + _IDX_SIZE, _IDX_SIZE)
+
+/** @brief Control block of secure packet buffer.
+ *
+ * The structure contains configuration data.
+ */
+struct pbuf_cfg {
+	volatile uint32_t *rd_idx_loc;	/* Address of the value holding first
+					 * valid byte in data[].
+					 */
+	volatile uint32_t *wr_idx_loc;	/* Address of the value holding first
+					 * free byte in data[].
+					 */
+	uint32_t cache_line_sz;		/* CPU data cache line size in bytes.
+					 * Used for validation - TODO: To be
+					 * replaced by flags.
+					 */
+	uint32_t len;			/* Length of data[] in bytes. */
+	uint8_t *data_loc;		/* Location of the buffer data. */
+};
+
+
+struct pbuf_data {
+	volatile uint32_t wr_idx;	/* Index of the first holding first
+					 * free byte in data[]. Used for
+					 * writing.
+					 */
+	volatile uint32_t rd_idx;	/* Index of the first holding first
+					 * valid byte in data[]. Used for
+					 * reading.
+					 */
+};
+
+
+/**
+ * @brief Scure packed buffer.
+ *
+ * The packet buffer implements lightweight unidirectional packet
+ * buffer with read/write semantics on top of a memory region shared
+ * by the reader and writer. It embeds cache and memory barrier management to
+ * ensure correct data access.
+ *
+ * This structure supports single writer and reader. Data stored in the buffer
+ * is encapsulated to a message (with length header). The read/write API is
+ * written in a way to protect the data from being corrupted.
+ */
+struct pbuf {
+	const struct pbuf_cfg *const cfg;	/* Configuration of the
+						 * buffer.
+						 */
+	struct pbuf_data data;			/* Data used to read and write
+						 * to the buffer
+						 */
+};
+
+/* Helper macro to check whenever the value is aligned. */
+#define IS_ALIGNED(address, alignment) \
```

Comment:
```suggestion
#define PBUF_IS_ALIGNED(address, alignment) \
```

Reply:
I have found macro in `include/toolchain/common.h` that fits perfectly for that, no need to define it yet, again. I used that macro instead.